### ISSUE TITLE
t/ffi_c_stat.t fails on FreeBSD systems

### ISSUE THREAD

**Issue by eserte** [uperl/FFI-C-Stat#5]
On newer FreeBSD systems (version 12, 13, 14) the test suite fails:
```
t/ffi_c_stat.t .. 1/?
# Failed test 'do a stat on a regular file'
# at t/ffi_c_stat.t line 30.
# +--------+------------------------+----+------------------------+--------+
# | PATH   | GOT                    | OP | CHECK                  | LNs    |
# +--------+------------------------+----+------------------------+--------+
# |        | FFI::C::Stat=SCALAR(0x |    | <OBJECT>               | 25, 29 |
# |        | 801bbe210)             |    |                        |        |
# |        |                        |    |                        |        |
# | dev()  | 12873038014975195776   | eq | -5573706058734355840   | 484    |
# | rdev() | 18446744073709551615   | eq | -1                     | 484    |
# +--------+------------------------+----+------------------------+--------+

# Failed test 'clone a stat'
# at t/ffi_c_stat.t line 40.
# +--------+------------------------+----+------------------------+--------+
# | PATH   | GOT                    | OP | CHECK                  | LNs    |
# +--------+------------------------+----+------------------------+--------+
# |        | FFI::C::Stat=SCALAR(0x |    | <OBJECT>               | 35, 39 |
# |        | 801bbe4c8)             |    |                        |        |
# |        |                        |    |                        |        |
# | dev()  | 12873038014975195776   | eq | -5573706058734355840   | 484    |
# | rdev() | 18446744073709551615   | eq | -1                     | 484    |
# +--------+------------------------+----+------------------------+--------+
... etc ...
```
However, I think FFI::C::Stat is correct here. Calling the `stat` utility on the test file outputs the following:
```
$ stat -r corpus/xx.txt
12873038014975195776 6165556 0100644 1 1023 1023 18446744073709551615 3 1622313704 1622155269 1622313704 1622155269 4096 9 2048 corpus/xx.txt
```
So my interpretation is that perl's `stat` is broken.

FreeBSD seems to use 64 bits for dev and rdev now. This changed either in version 11 or 12; on my FreeBSD 9 and 10 smokers the test suite passes, and I don't have 11 available.

**Comment by plicease** [comment]
Hrm.  I can't reproduce on my FreeBSD 12 VM.  Possibly by chance the dev / rdev fields fit in the smaller perl stat vars?

```
freebsd12% stat -r corpus/xx.txt 
59 4975948 0100644 1 1000 1000 9956735 3 1622319414 1622319414 1622319414 1622319414 32768 8 0 corpus/xx.txt
```

I guess we could we could work around this by getting the dev/rdev fields from the stat comment on FreeBSD (I probably wouldn't do it elsewhere for now, the output format doesn't seem to be consistent).  Do you know if this has been reported to p5p?

**Comment by plicease** [comment]
Seems like #6 would work around this, but the symlink test fails on the rdev check on my system:

```
not ok 5 - do a stat on a symlink

# Failed test 'do a stat on a symlink'
# at t/ffi_c_stat.t line 121.
# +--------+----------------------------------+----+----------+----------+
# | PATH   | GOT                              | OP | CHECK    | LNs      |
# +--------+----------------------------------+----+----------+----------+
# |        | FFI::C::Stat=SCALAR(0x8027d34f8) |    | <OBJECT> | 116, 120 |
# | rdev() | 8660267556691668835              | eq | 9956735  | 462      |
# +--------+----------------------------------+----+----------+----------+
1..5
```

My read of the documentation is that `stat -r` is actually an lstat and to get a lstat you use `stat -rL`?

**Comment by plicease** [comment]
Yeah so on my VM it looks like `FFI::C::Stat` matches the perl stat which is wrong :/

**Comment by eserte** [comment]
It's possible that dev and rdev are so large here because zfs is in use.

For perl's stat() I suspect that
```
	mPUSHi(PL_statcache.st_dev);
```
in `pp_stat` should use `mPUSHu` instead (at least on systems where dev_t is uint64).
I'll check tomorrow if this is the case, and prepare a p5p report.

**Comment by plicease** [comment]
I've confirmed that FFI::Platypus has defined `dev_t` as uint64 on my FreeBSD VM and that the C code in `ffi/stat.c` is getting the smaller rdev value on the symlink than what the command line `stat` is returning.  So I am perplexed by that.

@eserte Can you try #6 branch on your system?  (You should be able to run the tests with `prove` without needing dzil)

**Comment by eserte** [comment]
> So I am perplexed by that.

Maybe you used `printf("%ld",...)` instead of `printf("%lu",...)` for checking the value?

**Comment by plicease** [comment]
This should be addressed in 0.02.  Thanks for opening the ticket with perl.
